### PR TITLE
Linear warmup sched

### PR DESCRIPTION
**Pull request by ananyahjha93**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?   
- [x] Did you write any new necessary tests?  

## What does this PR do?
Fixes # (issue).

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pl_bolts/optimizers/lr_scheduler.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+        if self.last_epoch == 0:
+            return [self.warmup_start_lr] * len(self.base_lrs)
+        elif self.last_epoch < self.warmup_epochs:
+            return [group['lr'] + (base_lr - self.warmup_start_lr) / (self.warmup_epochs - 1)
+                    for base_lr, group in zip(self.base_lrs, self.optimizer.param_groups)]
+        elif self.last_epoch == self.warmup_epochs:
+            return self.base_lrs
+        # TODO: what does this do? This might be wrong for now since axis has shifted due to warmup
+        elif (self.last_epoch - 1 - self.max_epochs) % (2 * (self.max_epochs - self.warmup_epochs)) == 0:
```
Comment (ananyahjha93) [inline pl_bolts/optimizers/lr_scheduler.py:62]: @williamFalcon @Borda this is the T_curr = 2 (k + 1) T_max case for linear warmup + cosine annealing scheduler. This should fail but it doesn't. Only when you adjust it for the linear warmup part it fails. This elif condition is not adjusted for the linear warmup case.

File: tests/unit/optimizers/test_lr_scheduler.py [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+                assert abs(target[epoch] - param_group['lr']) < EPSILON,\
+                    "LR is wrong in epoch {}: expected {}, got {}".format(
+                        epoch, target[epoch], param_group['lr']
+                    )
+            [scheduler.step() for scheduler in schedulers]
+
+    def _test_against_closed_form(self, scheduler, closed_form_scheduler, epochs=10):
+        targets = []
+        for epoch in range(epochs):
+            closed_form_scheduler.step(epoch)
```
Comment (ananyahjha93) [inline tests/unit/optimizers/test_lr_scheduler.py:65]: this comes from earlier versions of pytorch where scheduler.step() was called before the epoch instead of after.

**Comment by ananyahjha93** [comment]
@Borda feedback on formatting as well, used https://github.com/psf/black on @nateraw 's advice.

File: pl_bolts/optimizers/lr_scheduler.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        ...     scheduler.step()
+        >>> #
+        >>> # passing epoch param case
+        >>> for epoch in range(40):
+        ...     scheduler.step(epoch)
+        ...     # train(...)
+        ...     # validate(...)
+    """
+
+    def __init__(
```
Comment (justusschock) [inline pl_bolts/optimizers/lr_scheduler.py:53]: can we add type annotations & docstrings here?

File: pl_bolts/optimizers/lr_scheduler.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    ):
+
+        self.warmup_epochs = warmup_epochs
+        self.max_epochs = max_epochs
+        self.warmup_start_lr = warmup_start_lr
+        self.eta_min = eta_min
+
+        super(LinearWarmupCosineAnnealingLR, self).__init__(optimizer, last_epoch)
+
+    def get_lr(self):
```
Comment (justusschock) [inline pl_bolts/optimizers/lr_scheduler.py:70]: Also return type annotations & docstrings :)

File: pl_bolts/optimizers/lr_scheduler.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+                    * (self.last_epoch - self.warmup_epochs - 1)
+                    / (self.max_epochs - self.warmup_epochs)
+                )
+            )
+            * (group["lr"] - self.eta_min)
+            + self.eta_min
+            for group in self.optimizer.param_groups
+        ]
+
+    def _get_closed_form_lr(self):
```
Comment (justusschock) [inline pl_bolts/optimizers/lr_scheduler.py:121]: Type annotations & docstrings as well :)

File: tests/unit/optimizers/test_lr_scheduler.py [hunk 6]
```diff
@@ -0,0 +1,434 @@
...
+from torch.optim.lr_scheduler import _LRScheduler
+
+import math
+import numpy as np
+from tests import reset_seed
+import pytorch_lightning as pl
+from pl_bolts.optimizers.lr_scheduler import LinearWarmupCosineAnnealingLR
+
+
+EPSILON = 1e-12
```
Comment (justusschock) [inline tests/unit/optimizers/test_lr_scheduler.py:13]: I think there is an epsilon defined somewhere in lightning iirc, maybe use this to avoid duplication?

cc @Borda who might know, where it is ^^

File: tests/unit/optimizers/test_lr_scheduler.py (hunk 2 above)
Comment (justusschock) [reply tests/unit/optimizers/test_lr_scheduler.py:76]: but do we still support these old versions?

Comment (ananyahjha93) [reply tests/unit/optimizers/test_lr_scheduler.py:76]: @justusschock current version of pytorch still defines ``` _test_against_closed_form()``` for now, all we need to do is define that function to support the old version. I have a warning and doc available for this

File: tests/unit/optimizers/test_lr_scheduler.py (hunk 6 above)
Comment (ananyahjha93) [reply tests/unit/optimizers/test_lr_scheduler.py:13]: @Borda so I had this question about default epsilons in lightning, ```numpy.finfo(numpy.float32).eps``` is 1.1920929e-07. Since learning rates used are up to 1e-6, we need a higher precision in schedulers. But, ```numpy.finfo(numpy.float64).eps``` comes out to be 2.220446049250313e-16, which results in failing tests. So the ideal epsilon seemed 1e-12.
